zrml/court/src/lib.rs

```diff
@@ -187,315 +169,1166 @@ mod pallet {
+                    if &lower_bound <= random_number && random_number < &upper_bound {
+                        let slashable = remainder.min(T::MinJurorStake::get());
+                        if let Some((weight, all_slashable)) = selections.get_mut(juror) {
+                            *weight = weight.saturating_add(1);
+                            *all_slashable = all_slashable.saturating_add(slashable);
                         } else {
-                            valid_winners_and_losers.push((ai, outcome));
+                            selections.insert(juror.clone(), (1, slashable));
                         }
+                        remainder = remainder.saturating_sub(slashable);
+                        random_set.remove(random_number);
+                    } else {
+                        break;
                     }
-                // Tardy juror didn't vote. Slash
-                } else if let JurorStatus::Tardy = juror.status {
-                    slash_and_remove_juror(ai)?;
                 }
-                // Ordinary juror didn't vote. Set as tardy
-                else {
-                    Self::set_stored_juror_as_tardy(ai)?;
+
+                if let Some((_, draw_slashable)) = selections.get_mut(juror) {
+                    *consumed_stake = consumed_stake.saturating_add(*draw_slashable);
+                    if let Some(mut juror_info) = <Jurors<T>>::get(&*juror) {
+                        juror_info.active_lock =
+                            juror_info.active_lock.saturating_add(*draw_slashable);
+                        <Jurors<T>>::insert(juror, juror_info);
+                    } else {
+                        debug_assert!(false, "Juror should exist in the Jurors map");
+                    }
+                }
+
+                if random_set.is_empty() {
+                    break;
                 }
+
+                current_weight = upper_bound;
             }
 
-            Ok(valid_winners_and_losers)
+            Ok(selections
+                .into_iter()
+                .map(|(juror, (weight, slashable))| Draw {
+                    juror,
+                    weight,
+                    vote: Vote::Drawn,
+                    slashable,
+                })
+                .collect())
         }
 
-        // Modifies a stored juror.
-        fn mutate_juror<F>(account_id: &T::AccountId, mut cb: F) -> DispatchResult
-        where
-            F: FnMut(&mut Juror) -> DispatchResult,
-        {
-            Jurors::<T>::try_mutate(account_id, |opt| {
-                if let Some(el) = opt {
-                    cb(el)?;
+        pub(crate) fn select_jurors(
+            market_id: &MarketIdOf<T>,
+            last_draws: DrawsOf<T>,
+            appeal_number: usize,
+        ) -> Result<DrawsOf<T>, DispatchError> {
+            let mut jurors: JurorPoolOf<T> = JurorPool::<T>::get();
+            let necessary_jurors_weight = Self::necessary_jurors_weight(appeal_number);
```

Comment:
I think you mentioned quite well that it should be associated to draws and not to jurors.

It's called because don't select `n` jurors, but `n * MinJurorStake` stake out of all jurors stakes.